### PR TITLE
Add OIDC Redis Token State Manager extension

### LINKED ISSUE #44382: Introduce `quarkus-oidc-redis-token-state-manager`
### Description

By default, Quarkus OIDC is stateless. It keeps tokens in the encrypted session cookies.

However, it is totally customizable with custom `TokenStateManager`. 
In fact, we ship `quarkus-oidc-db-token-state-manager`, nicely done by @michalvavrik, which makes it super easy to make Quarkus OIDC stateful and have tokens stored in a DB of choice.

However, shipping a quality cache based `TokenStateManager` out of the box can help some users with making a migration decision. For example, Vert.x supports Redis cache as well. Also, I'm aware of at least a single custom `TokenStateManager` which uses Redis. This is why I suggest to start with Redis and we can do Infinispan later as well. 
   

### Implementation ideas

Start with using Redis directly, similar to how we do it for storing and managing tokens in DB.
Then later, if Ladislav @Ladicek finds the time to complete his nice Vert.x based session support PR for Redis and Infinispan, we can just make `quarkus-oidc-redis-token-state-manager` call the Vertx Session API, and the whole extension will have a few lines of code only.

CC @michalvavrik @cescoffier

### PR DESCRIPTION
**Pull request by michalvavrik**
- closes: https://github.com/quarkusio/quarkus/issues/44382

### CODE REVIEW

File: extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            OidcRequestContext<Void> requestContext) {
+        return dataSource.key(String.class).del(tokenState).onFailure().recoverWithNull().replaceWithVoid();
+    }
+
+    private Uni<String> createTokenState(AuthorizationCodeTokensRecord tokenRecord, int attemptCount, SetArgs setArgs) {
+        if (attemptCount >= 3) {
+            return Uni.createFrom().failure(
+                    new RuntimeException("Failed to store OIDC token state in Redis as generated key already existed"));
+        }
+        final String key = "oidc-redis-token-state-" + Instant.now().getEpochSecond() + "-" + randomGenerator.nextInt();
```
Comment (cescoffier) [inline extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java:56]: I know it's weird, but... Redis use `:` as key separator.
So, I believe it should be something like `oidc:redis-token-state`

Also, any reason to use a time and a random id?

**Review by cescoffier — COMMENTED** [review]
I've added a small comment about the Redis key name.

Now, I'm wondering about the future. I was planning to extract Redis to a QV repository. I believe this will come with it (as well as the cache implementation). It's not planned for "now", but something for the future.

File: extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java (hunk 1 above)
Comment (Ladicek) [reply extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java:56]: AFAIK the `:` is only of historical significance, although `oidc:token` seems like a much better prefix than `oidc-redis-token-state`.

When it comes to using a timestamp and a random number, wouldn't a UUID be a much better choice?

**Review by Ladicek — APPROVED** [review]
LGTM, but see the question by @cescoffier

File: extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java (hunk 1 above)
Comment (michalvavrik) [reply extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java:56]: > So, I believe it should be something like oidc:redis-token-state
> AFAIK the : is only of historical significance, although oidc:token seems like a much better prefix than oidc-redis-token-state.

Alright, thank you, I didn't know about this. I think `oidc:token` is just fine. I'll update it.

> Also, any reason to use a time and a random id?

For random part see answer below. The time is just a cheap way to ensure that generated id doesn't already exist. I'd like to keep it.

> When it comes to using a timestamp and a random number, wouldn't a UUID be a much better choice?

`java.util.UUID#randomUUID` uses secure random. And the secure random was slower based on what I read. TBH I read it on Baeldung https://www.baeldung.com/java-17-random-number-generators#performance (see that table) so I don't know if that is true. I thought it's better to retry in that unlikely situation when random number already existed and use quicker algorithm.

Now, you could say that it attackers could guess random number if they know time range when this was requested and they got issues token state id before and after.

I suppose you would be right, I'll change it to `randomUUID`. thanks @Ladicek

Comment (Ladicek) [reply extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java:56]: I didn't even think of a possible attack vector, I don't understand security that well :-) I was just thinking that UUID exists exactly for this purpose, so why not use it. Performance would be a concern if this was called thousands time per second, but that shouldn't be the case here I think?

Comment (michalvavrik) [reply extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java]: > Performance would be a concern if this was called thousands time per second, but that shouldn't be the case here I think?

It shouldn't be frequent operation. I always think about performance about I didn't realize compelling reasons for using secure random. I've update it, please check if you want.

File: extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                });
+    }
+
+    private static String createTokenKey(String tokenState) {
+        return REDIS_KEY_PREFIX + tokenState;
+    }
+
+    private static String toTokenKey(String tokenState) {
+        return REDIS_KEY_PREFIX + tokenState;
+    }
```
[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java]: sorry, I was in hurry, it's gone now

File: extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+            OidcRequestContext<Void> requestContext) {
+        return dataSource.key(String.class).del(toTokenKey(tokenState)).onFailure().recoverWithNull().replaceWithVoid();
+    }
+
+    private Uni<String> createTokenState(AuthorizationCodeTokensRecord tokens, int attemptCount, SetArgs setArgs) {
+        if (attemptCount >= 3) {
+            return Uni.createFrom().failure(
+                    new RuntimeException("Failed to store OIDC token state in Redis as generated key already existed"));
+        }
+        final String tokenState = "" + Instant.now().getEpochSecond() + UUID.randomUUID();
```
[the remark this answers is not in the document]
Comment (michalvavrik) [reply extensions/oidc-redis-token-state-manager/runtime/src/main/java/io/quarkus/oidc/redis/token/state/manager/runtime/OidcRedisTokenStateManager.java]: Sure, there is no risk so I have dropped it. Thanks

**Review by Ladicek — APPROVED** [review]

**Review by sberyozkin — APPROVED** [review]
Thanks @michalvavrik, very neat and simple, thanks @Ladicek for the review

**Comment by sberyozkin** [comment]
@calvernaz, FYI

**Comment by sberyozkin** [comment]
Hi @michalvavrik @cescoffier
Apologies I missed the conversation about possible moving of Quarkus Redis Client to Quarkiverse. Yeah, we'll need to decide if the OIDC extension which depends on Redis Client will have to follow it and go to Quarkiverse itself, or may be, we should use this opportunity and figure out how Quarkus externsions living in the main repository can work together with Quarkiverse extensions... Otherwise, if this extension also moves, it is unclear if we have it supported at the product level... It will be an interesting discussion :-)

**Pull request merged**